tools/codegen/gen_vmap_plumbing2.py

```diff
@@ -0,0 +1,213 @@
+            )
+            idx += 2
+        else:
+            wrapped_returns.append(f'std::get<{idx}>(results)')
+            idx += 1
+    if len(wrapped_returns) == 1:
+        wrapped_returns = f'return {wrapped_returns[0]};'
+    else:
+        wrapped_returns = f'return std::make_tuple({", ".join(wrapped_returns)});'
+    return wrapped_returns
+
+
+def accepts_at_least_one_tensor_input(schema):
+    for arg in schema.arguments.flat_all:
+        if arg.type.is_tensor_like():
+            return True
+    return False
+
+
+def gen_vmap_inplace_plumbing(native_function):
+    schema = native_function.func
+    sig = DispatcherSignature.from_schema(schema)
+    returns = schema.returns
+
+    assert schema.kind() == SchemaKind.inplace
+
+    # Only support cases where all returns are Tensors or vector<Tensor>
+    if len(returns) == 0:
+        return None
+    if not all(is_tensor(ret.type) or is_vector_tensor(ret.type) for ret in returns):
+        return None
+    if not accepts_at_least_one_tensor_input(schema):
+        return None
+
+    unwraps, unwrapped_arg_list = gen_unwraps(schema.arguments.flat_all)
+    bdims_all_none_case = gen_case_where_all_bdims_are_none(schema.arguments.flat_all, schema)
+
+    return f"""\
+template <typename batch_rule_t, batch_rule_t batch_rule>
+{sig.decl(name=schema.name.unambiguous_name() + '_generated_plumbing')} {{
+  c10::impl::ExcludeDispatchKeyGuard guard(kBatchedKey);
+  auto maybe_layer = maybeCurrentDynamicLayer();
+  TORCH_INTERNAL_ASSERT(maybe_layer.has_value());
+  int64_t cur_level = maybe_layer->layerId();
+{textwrap.indent(bdims_all_none_case, "  ")}
+{textwrap.indent(unwraps, "  ")}
+  batch_rule({', '.join(unwrapped_arg_list)});
+  return {schema.arguments.flat_all[0].name};
+}}"""
+
+
+def gen_vmap_plumbing(native_function: NativeFunction) -> str:
+    schema = native_function.func
+    sig = DispatcherSignature.from_schema(schema)
+    returns = schema.returns
+
+    # Only support cases where all returns are Tensors or vector<Tensor>
+    if len(returns) == 0:
+        return None
+    if not all(is_tensor(ret.type) or is_vector_tensor(ret.type) for ret in returns):
+        return None
+    if not accepts_at_least_one_tensor_input(schema):
+        return None
+    # in-place views need special handling
+    if native_function.tag == Tag.inplace_view:
+        return None
+
+    if schema.kind() == SchemaKind.inplace:
+        return gen_vmap_inplace_plumbing(native_function)
+
+    # Don't support these
+    if schema.kind() == SchemaKind.out:
+        return None
+
+    unwraps, unwrapped_arg_list = gen_unwraps(schema.arguments.flat_all)
+    bdims_all_none_case = gen_case_where_all_bdims_are_none(schema.arguments.flat_all, schema)
+
+    wrapped_returns = gen_returns(returns)
+    return f"""\
+template <typename batch_rule_t, batch_rule_t batch_rule>
+{sig.decl(name=schema.name.unambiguous_name() + '_generated_plumbing')} {{
```

Comment:
I'm curious what the interface between the handwritten code and the codegen is, since it looks like you're generating a template (and this codegen doesn't produce any dispatcher registrations).

Where do we end up calling this template and/or registering stuff to the dispatcher?

Reply:
Aha! Great question.

We register things to the dispatcher with a VMAP_SUPPORT macro: https://github.com/pytorch/functorch/blob/8acf2d1663505f1c593baf6e22de5e34d8f69915/functorch/csrc/BatchRulesViews.cpp#L501-L502

VMAP_SUPPORT is implemented as follows: https://github.com/pytorch/functorch/blob/main/functorch/csrc/BatchRulesHelper.h#L52-L56

Essentially the templates may or may not get used:
- a user makes a registration using e.g. VMAP_SUPPORT(diag, diag_batch_rule);
- This passes diag_batch_rule to the generated template and registers it to the batched key. 

The template is there because we don't have something like derivatives.yaml that tells us ahead-of-time what functions to register to the key, instead the rules are dynamically registered. Maybe we'll change this sometime.

Reply:
> The template is there because we don't have something like derivatives.yaml that tells us ahead-of-time what functions to register to the key, instead the rules are dynamically registered. Maybe we'll change this sometime.

Reply:
@bdhirsh did mean to comment something?

Reply:
whoops 😛 

Yep, everything you said totally makes sense (and I think this PR is good to land whenever you think it's cleaned up). `VMAP_SUPPORT` requires less-boilerplate to add new ops compared to a yaml file (where you need to both edit the yaml, and write a kernel with the "right" signature), which seems pretty nice. I'm mostly curious how the codegen is gonna end up being extended over time, but that's probably a conversation for later (once we have tags).